engineering-plans/0005-subgraph-composition-query-execution.md

```diff
@@ -0,0 +1,123 @@
+# PLAN-0005: Subgraph Composition Query Execution
+
+<dl>
+  <dt>Author</dt>
+  <dd>Jorge Olivero</dd>
+
+  <dt>Implements</dt>
+  <dd><a href="../rfcs/0001-subgraph-composition.md">RFC-0001 Subgraph Composition</a></dd>
+
+  <dt>Engineering Plan pull request</dt>
+  <dd><a href="">Engineering Plan PR</a></dd>
+
+  <dt>Obsoletes (if applicable)</dt>
+  <dd>-</dd>
+
+  <dt>Date of submission</dt>
+  <dd>2020-1-14</dd>
+
+  <dt>Date of approval</dt>
+  <dd>TBD</dd>
+
+  <dt>Approved by</dt>
+  <dd>TBD</dd>
+</dl>
+
+## Summary
+
+Since subgraph composition allows a subgraph to import types from another subgraph, GraphQL query execution will need to traverse a subgraph's import graph to resolve concrete types.
+Because imported schemas/types are loosely coupled, query execution also needs to address missing types in the api schema during resolution.
+In addition, imported types which have been renamed need to be mapped back to their original name so that queries can resolve correctly.
+Lastly, to simplify query execution graph-node needs to deny subgraphs with JSONB storage from using subgraph composition.
+
+## Implementation
+
+1. Support Custom Type Names: Ensure @originalName directives on the imported types are used to query the correct sql tables.
+2. Codify New Query Runtime Errors: Ensure @placeholder types are acknowledged and proper errors are returned during query execution when a type's subgraph is not deployed to the graph-node resolving the query.
+3. Query Correct Subgraph: Ensure that @subgraphId directives on imported types are used correctly in all cases.
+4. Add Subgraph Composition Feature Flag: Ensure subgraphs JSONB storage can not use subgraph composition.
+
+### Support Custom Type Names
+
+When a type is imported with the name/as syntax:
+
+```graphql
+type _Schema_ @import(types: [{ name: "A", as: "B" }], from: { id: "..." })
+```
+
+an `@originalName` directive is added to it when it is merged into the target schema:
+
+```graphql
+type B @entity @subgraphId(id: "...") @originalName(name: "A") {
+	id: ID!
+	...
+}
+```
+
+GraphQL query resolution needs to make use of the `@originalName` directive to ensure SQL queries are made against the correct tables.
+
+The follow code locations need to be updated:
+
+1. https://github.com/graphprotocol/graph-node/blob/master/graphql/src/store/resolver.rs#L362-L379f
+
+2. https://github.com/graphprotocol/graph-node/blob/master/graphql/src/store/query.rs#L21-L27
```

Comment:
What's very important in engineering plans is that we not just know _where_ something needs to be changed but that we have clarity about _how_ things have to change. This is crucial to assessing the risk of the plan (with regards to feasibility and estimates) and is missing here.